Standards/scs-0215-v1-robustness-features.md

```diff
@@ -0,0 +1,267 @@
+Possible flags, that can be set for this feature, are:
+
+* auto-compaction-mode
+* auto-compaction-retention
+
+etcd cluster defragmentation unfortunately can't be done automatically. Instead the user would need to manually call
+the defrag command on the cluster. In order to mitigate this, a systemd (or similar) job could be created, which
+periodically calls the defragmentation procedure.
+An example for such a job and its helpers could be the following:
+
+```bash
+[Unit]
+Description=Run etcdctl defrag
+Documentation=https://etcd.io/docs/v3.3.12/op-guide/maintenance/#defragmentation
+After=network.target
+[Service]
+Type=oneshot
+Environment="LOG_DIR=/var/log"
+Environment="ETCDCTL_API=3"
+ExecStart=/usr/local/sbin/etcdctl --cert /etc/kubernetes/pki/etcd/peer.crt --key /etc/kubernetes/pki/etcd/peer.key --cacert /etc/kubernetes/pki/etcd/ca.crt defrag
+[Install]
+WantedBy=multi-user.target
+```
+
+```bash
+[Unit]
+Description=Run etcd-defrag.service every day
+After=network.target
+[Timer]
+OnCalendar=*-*-* 02:00:0
+RandomizedDelaySec=10m
+[Install]
+WantedBy=multi-user.target
+```
+
+More information about both of these procedures can be found in the respective etcd documentation
+[etcd maintenance](https://etcd.io/docs/v3.3/op-guide/maintenance/)
+
+### etcd backup
+
+An etcd cluster should be regularly backed up in order to be able to restore the cluster to a known good state at an
+earlier space in time if a failure or incorrect state happens.
+The cluster should be backed up multiple times in order to have different possible states to go back to. This is especially
+useful, if data in the newer backups was already corrupted in some way or important data was deleted in them.
+For this reason, a backup strategy needs to be developed with a decreasing number of backups in an increasing period of time,
+meaning that the previous year should only have 1 backup, but the current week should have multiple.
+Information about the backup process can be found in the etcd documentation:
+[Upgrade etcd](https://kubernetes.io/docs/tasks/administer-cluster/configure-upgrade-etcd/)
+
+### CA expiration avoidance
+
+In order to secure the communication of a Kubernetes cluster, (TLS) certificates signed by a controlled
+Certificate Authority (CA) can be used.
+Normally, these certificates expire after a set period of time. In order to avoid expiration and failure of a cluster,
+these certificates need to be rotated regularly and at best automatically.
+It is important to either set `--rotate-server-certificates` as a command line parameter or set `rotateCertificates: true`
+in the kubelet config or the `kubeletExtraArgs` of the cluster-template.yaml file. This activates the rotation of the
+kubelet server certificate. Another recommendation is to set `serverTLSBootstrap: true`, which also enables the request
+and rotation of the certificate for the kubelet according to the documentation
```

Comment:
IMO I find the explanation regarding the rotation of kubelet server certificates somewhat unclear. In addition, according to the documentation, it appears that the approval of Certificate Signing Requests (CSRs) should be done either manually or by using a custom controller, as indicated in this note [here](https://kubernetes.io/docs/reference/access-authn-authz/kubelet-tls-bootstrapping/#certificate-rotation).

Overall, I do not see the recommendation to set the kubelet parameters to automate server certificate rotation as a complete solution because it is just a part of Kubernetes [PKI](https://kubernetes.io/docs/setup/best-practices/certificates/). 

Instead, I would prefer 2 approaches you described below, i.e. 

1. Update the cluster at least once a year to rotate certificates automatically
2.  Renew all certificates with `kubeadm certs renew all`
     - You only need to do this when you don't upgrade your cluster
     
the above is also in sync with SCS KaaS reference implementation, see [docs](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/blob/main/doc/Maintenance_and_Troubleshooting.md#client-certificates-in-kubernetes-expire-after-one-year)

Reply:
I probably wrote this section a bit unclear then and I also probably understand something wrong in the certificate rotation documentation (and some other documents).
I planned this section to basically recommend the cluster renewal or certificate renewal approaches you described. But I was under the impression, that the flags described in the first subsection are required to have all the certificates exchanged correctly.

Reply:
> But I was under the impression, that the flags described in the first subsection are required to have all the certificates exchanged correctly.

After rereading the docs (again), you are right. Kubeadm does not manage the kubelet cert rotation. 
Based on this [docs](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubelet-serving-certs), my understanding is as follows:

1. The command `kubeadm certs renew all` will renew Kubernetes control plane certificates, but it won't renew the certificates used for authenticating the kubelet against the k8s API server and won't renew the kubelet server certificate
2. The kubelet parameter `rotateCertificates` (defaults to true) will enable the full automatic renewal of kubelet client certificate for authenticating the kubelet against the k8s API
3.  The kubelet parameter  `TLSBootstrap: true` will enable the bootstrap of kubelet server certificates by requesting them from the certificates.k8s.io API. 
 ! One known limitation is that the CSRs (Certificate Signing Requests) for these certificates cannot be automatically approved. This will require manual action from the user or a third-party controller like [kubelet-csr-approver](https://github.com/postfinance/kubelet-csr-approver)

Reply:
I updated the section a bit more to include the possibility for a CSR approver controller and also a better representation for the `kubeadm certs renew all` command.